src/nvim/cursor_shape.c

```diff
@@ -7,40 +7,84 @@
+#include "nvim/ui.h"
 
-/*
- * Handling of cursor and mouse pointer shapes in various modes.
- */
-
+/// Handling of cursor and mouse pointer shapes in various modes.
 static cursorentry_T shape_table[SHAPE_IDX_COUNT] =
 {
-  /* The values will be filled in from the 'guicursor' and 'mouseshape'
-   * defaults when Vim starts.
-   * Adjust the SHAPE_IDX_ defines when making changes! */
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "n", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "v", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "i", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "r", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "c", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "ci", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "cr", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "o", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0, 700L, 400L, 250L, 0, 0, "ve", SHAPE_CURSOR+SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "e", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "s", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "sd", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "vs", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "vd", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "m", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "ml", SHAPE_MOUSE},
-  {0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR},
+  // The values will be filled in from the 'guicursor' and 'mouseshape'
+  // defaults when Vim starts.
+  // Adjust the SHAPE_IDX_ defines when making changes!
+  { "normal",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "n", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "visual",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "v", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "insert",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "i", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "replace",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "r", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "cmd_normal",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "c", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "cmd_insert", 0,
+    0, 0, 700L, 400L, 250L, 0, 0, "ci", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "cmd_replace",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "cr", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "pending",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "o", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "visual_select",
+    0, 0, 0, 700L, 400L, 250L, 0, 0, "ve", SHAPE_CURSOR+SHAPE_MOUSE },
+  { "cmd_line", 0, 0, 0,   0L,   0L,   0L, 0, 0, "e", SHAPE_MOUSE },
+  { "statusline", 0, 0, 0,   0L,   0L,   0L, 0, 0, "s", SHAPE_MOUSE },
+  { "drag_statusline", 0, 0, 0,   0L,   0L,   0L, 0, 0, "sd", SHAPE_MOUSE },
+  { "vsep", 0, 0, 0,   0L,   0L,   0L, 0, 0, "vs", SHAPE_MOUSE },
+  { "vdrag", 0, 0, 0,   0L,   0L,   0L, 0, 0, "vd", SHAPE_MOUSE },
+  { "more", 0, 0, 0,   0L,   0L,   0L, 0, 0, "m", SHAPE_MOUSE },
+  { "more_lastline", 0, 0, 0,   0L,   0L,   0L, 0, 0, "ml", SHAPE_MOUSE },
+  { "match_paren", 0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR },
```

Comment:
You must be right I just joined the comment as a description `#define SHAPE_IDX_SM    16      /* showing matching paren */. I was not aware of the difference